### ISSUE TITLE
Timeout raises exception

### ISSUE THREAD

**Issue by mohmad-null** [scrapy/scrapy#2442]
This issue has several components, all related to TimeoutErrors.
Scrapy 1.2.1

a) If a TimeoutError is raised, by default it will print the entire exception to the logger. This isn't in keeping with other exceptions (DNSLookupError, TCPTimedOutError, ConnectError) which are all transparent.

~~b) The TimeoutError trace without a errback will halt the processing of the spider. This isn't in keeping with other exceptions (DNSLookupError, TCPTimedOutError, ConnectError) which are all transparent.~~

**Comment by redapple** [comment]
Can you provide sample code reproducing this?

**Comment by mohmad-null** [comment]
Item (b) is wrong. It just looked like it had crashed it because it was waiting for the timeout.

Item (a) is still valid. You can replicate it with any request to a service that you control and suspend that service (suspend the "process") so it times out.

**Comment by redapple** [comment]
@mohmad-null , can you help us by providing some test code reproducing the issue (ideally with test server and test spider)?
There are at least 2 `"TimeoutError"` used and checked throughout scrapy code base: `twisted.internet.error.TimeoutError` and `twisted.internet.defer.TimeoutError`.
Which one are you referring to?

> ... other exceptions (DNSLookupError, TCPTimedOutError, ConnectError) which are all transparent.

what do you mean by "transparent"?

See below where I tested 3 types of timeout and I can see the exception written to the console.
Are you describing something different?
Does it depend on the logging destination (console vs. file perhaps)?

```
(scrapy12) paul@paul-SATELLITE-R830:~$ scrapy shell http://www.example.com -s DOWNLOAD_TIMEOUT=0.01
2016-12-13 10:43:03 [scrapy] INFO: Scrapy 1.2.2 started (bot: scrapybot)
2016-12-13 10:43:03 [scrapy] INFO: Overridden settings: {'DOWNLOAD_TIMEOUT': '0.01', 'LOGSTATS_INTERVAL': 0, 'DUPEFILTER_CLASS': 'scrapy.dupefilters.BaseDupeFilter'}
(...)
2016-12-13 10:43:03 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-12-13 10:43:03 [scrapy] INFO: Spider opened
2016-12-13 10:43:03 [scrapy] DEBUG: Retrying <GET http://www.example.com> (failed 1 times): User timeout caused connection failure.
2016-12-13 10:43:03 [scrapy] DEBUG: Retrying <GET http://www.example.com> (failed 2 times): User timeout caused connection failure.
2016-12-13 10:43:03 [scrapy] DEBUG: Gave up retrying <GET http://www.example.com> (failed 3 times): User timeout caused connection failure.
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapy12/bin/scrapy", line 11, in <module>
    sys.exit(execute())
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 142, in execute
    _run_print_help(parser, _run_command, cmd, args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 88, in _run_print_help
    func(*a, **kw)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 149, in _run_command
    cmd.run(args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/commands/shell.py", line 71, in run
    shell.start(url=url)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 47, in start
    self.fetch(url, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 112, in fetch
    reactor, self._schedule, request, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/twisted/internet/threads.py", line 122, in blockingCallFromThread
    result.raiseException()
  File "<string>", line 2, in raiseException
twisted.internet.error.TimeoutError: User timeout caused connection failure.


(scrapy12) paul@paul-SATELLITE-R830:~$ scrapy shell http://www.exampleeeeeeeeeeeeeee.com 
2016-12-13 10:43:52 [scrapy] INFO: Scrapy 1.2.2 started (bot: scrapybot)
2016-12-13 10:43:52 [scrapy] INFO: Overridden settings: {'LOGSTATS_INTERVAL': 0, 'DUPEFILTER_CLASS': 'scrapy.dupefilters.BaseDupeFilter'}
(...)
2016-12-13 10:43:52 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-12-13 10:43:52 [scrapy] INFO: Spider opened
2016-12-13 10:43:52 [scrapy] DEBUG: Retrying <GET http://www.exampleeeeeeeeeeeeeee.com> (failed 1 times): DNS lookup failed: address 'www.exampleeeeeeeeeeeeeee.com' not found: [Errno -5] No address associated with hostname.
2016-12-13 10:43:52 [scrapy] DEBUG: Retrying <GET http://www.exampleeeeeeeeeeeeeee.com> (failed 2 times): DNS lookup failed: address 'www.exampleeeeeeeeeeeeeee.com' not found: [Errno -5] No address associated with hostname.
2016-12-13 10:43:52 [scrapy] DEBUG: Gave up retrying <GET http://www.exampleeeeeeeeeeeeeee.com> (failed 3 times): DNS lookup failed: address 'www.exampleeeeeeeeeeeeeee.com' not found: [Errno -5] No address associated with hostname.
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapy12/bin/scrapy", line 11, in <module>
    sys.exit(execute())
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 142, in execute
    _run_print_help(parser, _run_command, cmd, args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 88, in _run_print_help
    func(*a, **kw)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 149, in _run_command
    cmd.run(args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/commands/shell.py", line 71, in run
    shell.start(url=url)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 47, in start
    self.fetch(url, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 112, in fetch
    reactor, self._schedule, request, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/twisted/internet/threads.py", line 122, in blockingCallFromThread
    result.raiseException()
  File "<string>", line 2, in raiseException
twisted.internet.error.DNSLookupError: DNS lookup failed: address 'www.exampleeeeeeeeeeeeeee.com' not found: [Errno -5] No address associated with hostname.


(scrapy12) paul@paul-SATELLITE-R830:~$ scrapy shell http://www.example.com:8080
2016-12-13 10:44:32 [scrapy] INFO: Scrapy 1.2.2 started (bot: scrapybot)
(...)
2016-12-13 10:44:32 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-12-13 10:44:32 [scrapy] INFO: Spider opened
2016-12-13 10:46:39 [scrapy] DEBUG: Retrying <GET http://www.example.com:8080> (failed 1 times): TCP connection timed out: 110: Connection timed out.
2016-12-13 10:48:47 [scrapy] DEBUG: Retrying <GET http://www.example.com:8080> (failed 2 times): TCP connection timed out: 110: Connection timed out.
2016-12-13 10:50:54 [scrapy] DEBUG: Gave up retrying <GET http://www.example.com:8080> (failed 3 times): TCP connection timed out: 110: Connection timed out.
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapy12/bin/scrapy", line 11, in <module>
    sys.exit(execute())
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 142, in execute
    _run_print_help(parser, _run_command, cmd, args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 88, in _run_print_help
    func(*a, **kw)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/cmdline.py", line 149, in _run_command
    cmd.run(args, opts)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/commands/shell.py", line 71, in run
    shell.start(url=url)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 47, in start
    self.fetch(url, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/scrapy/shell.py", line 112, in fetch
    reactor, self._schedule, request, spider)
  File "/home/paul/.virtualenvs/scrapy12/local/lib/python2.7/site-packages/twisted/internet/threads.py", line 122, in blockingCallFromThread
    result.raiseException()
  File "<string>", line 2, in raiseException
twisted.internet.error.TCPTimedOutError: TCP connection timed out: 110: Connection timed out.
```

**Comment by mohmad-null** [comment]
This is twisted.internet.error.TimeoutError  - I've never seen the "defer" one to which you refer.

Replicating is easy.
Command line:

```
> scrapy startproject tutorial
> scrapy genspider example example.com
```

Edit the example.py so that it points to a server your control (and ensure it's in the allowed_domains of course):
`    start_urls = ['http://localhost/my_test.html']`

run it:

`> scrapy crawl example`

It will run fine.
Now suspend the webserver (i'm using apache on windows - using process explorer I right click and "suspend" the process - I'm assuming it's easily done on Linux too).

Run again. This time after 9 mins and a number of DEBUG level events, it will raise the exception:

> Getting http://localhost:8080/robots.txt took longer than 180.0 seconds..
> 2016-12-14 20:38:00 [scrapy] ERROR: Error downloading <GET http://localhost:8080/robots.txt>: User timeout caused connection failure: Getting http://l
> ocalhost:8080/robots.txt took longer than 180.0 seconds..
> Traceback (most recent call last):
>   File "c:\python27_internet\lib\site-packages\twisted\internet\defer.py", line 1183, in _inlineCallbacks
>     result = result.throwExceptionIntoGenerator(g)
>   File "c:\python27_internet\lib\site-packages\twisted\python\failure.py", line 389, in throwExceptionIntoGenerator
>     return g.throw(self.type, self.value, self.tb)
>   File "c:\python27_internet\lib\site-packages\scrapy\core\downloader\middleware.py", line 43, in process_request
>     defer.returnValue((yield download_func(request=request,spider=spider)))
>   File "c:\python27_internet\lib\site-packages\twisted\internet\defer.py", line 587, in _runCallbacks
>     current.result = callback(current.result, *args, **kw)
>   File "c:\python27_internet\lib\site-packages\scrapy\core\downloader\handlers\http11.py", line 305, in _cb_timeout
>     raise TimeoutError("Getting %s took longer than %s seconds." % (url, timeout))
> TimeoutError: User timeout caused connection failure: Getting http://localhost:8080/robots.txt took longer than 180.0 seconds..


Now, edit example.py to point to a made up start_url:

`    start_urls = ['http://complete mady up domain/test.html']`

Run it and it will fail the `DNSLookupError`. However, it's "_transparent_" in that there isn't a full exception. There's a single ERROR level event and three DEBUG level events describing the problem.

Basically, I don't think the Timeout should be raising an exception, it should be doing the same as every other problem (such as the DNSLookupError induced above) - logging an appropriate ERROR level event.

I hope that's clearer. I've never seen a stacktrace for a DNSLookupError myself.

**Comment by hanjihun** [comment]
When I specify the DOWNLOAD_TIMEOUT value in the settings.py, it occurs.

**Comment by echoocking** [comment]
me too

**Comment by echoocking** [comment]
but it not always occur .
in my spider 
During the program is running， sometimes is fine。the log is 
2017-05-27 14:23:12 [crawler.download_middlewares.RetryMiddleware] DEBUG: Gave up retrying <GET ***url ***(failed 3 times): User timeout caused connection failure: Getting https://p.3.cn/prices/mgets?callback=jQuery5474916&type=1&skuIds=J_12157389320&source=item-pc&pdtk=&pduid=1491547425047318305720 took longer than 5.0 seconds..


but to deal with another request, it will show 
Traceback (most recent call last):
  File "/Users/echoocking/workspace/scrapy/lib/python2.7/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/Users/echoocking/workspace/scrapy/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/Users/echoocking/workspace/scrapy/lib/python2.7/site-packages/scrapy/core/downloader/middleware.py", line 43, in process_request
    defer.returnValue((yield download_func(request=request,spider=spider)))
  File "/Users/echoocking/workspace/scrapy/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Users/echoocking/workspace/scrapy/lib/python2.7/site-packages/scrapy/core/downloader/handlers/http11.py", line 306, in _cb_timeout
    raise TimeoutError("Getting %s took longer than %s seconds." % (url, timeout))
TimeoutError: User timeout caused connection failure: Getting http://*****.html took longer than 5.0 seconds..

**Comment by TanJay** [comment]
I also have the same issue. I tried this site with python request/urllib2 as well same result with timeout.

When I use scrapy I get,
```
MacBook-Pro:~ tanushajayasinghe$ scrapy shell https://www.abercrombie.com/shop/wd/mens-underwear-and-socks
:0: UserWarning: You do not have a working installation of the service_identity module: 'No module named cryptography.x509'.  Please install it from <https://pypi.python.org/pypi/service_identity> and make sure all of its dependencies are satisfied.  Without the service_identity module and a recent enough pyOpenSSL to support it, Twisted can perform only rudimentary TLS client hostname verification.  Many valid certificate/hostname mappings may be rejected.
2017-12-16 19:10:45 [scrapy.utils.log] INFO: Scrapy 1.4.0 started (bot: scrapybot)
2017-12-16 19:10:45 [scrapy.utils.log] INFO: Overridden settings: {'LOGSTATS_INTERVAL': 0, 'DUPEFILTER_CLASS': 'scrapy.dupefilters.BaseDupeFilter'}
2017-12-16 19:10:45 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.corestats.CoreStats']
2017-12-16 19:10:45 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2017-12-16 19:10:45 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2017-12-16 19:10:45 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2017-12-16 19:10:45 [scrapy.extensions.telnet] DEBUG: Telnet console listening on 127.0.0.1:6024
2017-12-16 19:10:45 [scrapy.core.engine] INFO: Spider opened



2017-12-16 19:13:45 [scrapy.downloadermiddlewares.retry] DEBUG: Retrying <GET https://www.abercrombie.com/shop/wd/mens-underwear-and-socks> (failed 1 times): User timeout caused connection failure: Getting https://www.abercrombie.com/shop/wd/mens-underwear-and-socks took longer than 180.0 seconds..
2017-12-16 19:16:45 [scrapy.downloadermiddlewares.retry] DEBUG: Retrying <GET https://www.abercrombie.com/shop/wd/mens-underwear-and-socks> (failed 2 times): User timeout caused connection failure: Getting https://www.abercrombie.com/shop/wd/mens-underwear-and-socks took longer than 180.0 seconds..
```

**Comment by wRAR** [comment]
> I've never seen a stacktrace for a DNSLookupError myself.

```
2025-01-06 19:51:02 [scrapy.downloadermiddlewares.retry] ERROR: Gave up retrying <GET http://vdfvdfvdf.dcs> (failed 3 times): DNS lookup failed: no results for hostname lookup: vdfvdfvdf.dcs.
2025-01-06 19:51:02 [scrapy.core.scraper] ERROR: Error downloading <GET http://vdfvdfvdf.dcs>
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 2013, in _inlineCallbacks
    result = context.run(
             ^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in throwExceptionIntoGenerator
    return g.throw(self.value.with_traceback(self.tb))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/scrapy/core/downloader/middleware.py", line 68, in process_request
    return (yield download_func(request, spider))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1088, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/twisted/internet/endpoints.py", line 1061, in startConnectionAttempts
    raise error.DNSLookupError(
twisted.internet.error.DNSLookupError: DNS lookup failed: no results for hostname lookup: vdfvdfvdf.dcs.
```

I'm closing this because the behavior is consistent for different exceptions unlike what you said.
If this is still now what you've meant you can try explaining again with more details.
